### ISSUE TITLE
Rename Symbol deletes extra characters when replacing a symbol found after an elif

### ISSUE THREAD

**Issue by SmartLamScott** [microsoft/pylance-release#4966]
Type: <b>Bug</b>

# Behaviour
## Expected vs. Actual

When using "Rename Symbol" to rename a variable found on the same line as an elif statement, the preceding whitespace and surrounding parenthesis (if present) are erronously deleted. They should remain untouched.

## Steps to reproduce:

1. Create an elif statement that uses a variable as its only argument.
2. Highlight the variable (from somewhere other than the line with the elif.
3. Run the command "Rename Symbol" and rename the variable to anything else.
4. All characters except whitespace before "elif", "elif", the new variable name, and the trailing colon will be deleted. This includes the whitespace between "elif" and the new variable name, as well as any parenthesis that may be present on the line.

# Diagnostic data

-   Python version (& distribution if applicable, e.g. Anaconda): 3.11.5
-   Type of virtual environment used (e.g. conda, venv, virtualenv, etc.): Global
-   Value of the `python.languageServer` setting: Default

<details>

<summary>Output for <code>Python</code> in the <code>Output</code> panel (<code>View</code>→<code>Output</code>, change the drop-down the upper-right of the <code>Output</code> panel to <code>Python</code>)
</summary>

<p>

```
XXX
```

</p>
</details>

<details>

<summary>User Settings</summary>

<p>

```

languageServer: "Pylance"

```

</p>
</details>

Extension version: 2023.18.0
VS Code version: Code 1.83.1 (f1b07bd25dfad64b0167beb15359ae573aecd2cc, 2023-10-10T23:48:05.904Z)
OS version: Windows_NT x64 10.0.22621
Modes:

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|13th Gen Intel(R) Core(TM) i7-13700H (20 x 2918)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: enabled_on<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: enabled|
|Load (avg)|undefined|
|Memory (System)|15.70GB (2.78GB free)|
|Process Argv|--crash-reporter-id 0b9ddb31-2fef-4557-bf98-877efdbe1c71|
|Screen Reader|no|
|VM|0%|
</details><details>
<summary>A/B Experiments</summary>

```
vsliv368:30146709
vsreu685:30147344
python383:30185418
vspor879:30202332
vspor708:30202333
vspor363:30204092
vswsl492cf:30256860
vslsvsres303:30308271
vserr242cf:30382550
pythontb:30283811
vsjup518:30340749
pythonptprofiler:30281270
vshan820:30294714
vstes263:30335439
vscod805:30301674
binariesv615:30325510
bridge0708:30335490
bridge0723:30353136
vsaa593:30376534
pythonvs932:30410667
py29gd2263:30856252
vscaac:30438847
vsclangdf:30486550
c4g48928:30535728
dsvsc012:30540252
pynewext54:30695312
azure-dev_surveyone:30548225
2e4cg342:30602488
89544117:30613380
showlangstatbar:30737416
962ge761:30841074
03d35959:30757346
pythonfmttext:30731395
f8hc8238:30694864
fixshowwlkth:30771522
showindicator:30805244
pythongtdpath:30769146
i26e3531:30792625
pythonnosmt12:30797651
pythonidxpt:30805730
pythonnoceb:30805159
copilotsettingt:30859503
dsvsc013:30795093
dsvsc014:30804076
diffeditorv2:30821572
dsvsc015:30845448
pythonregdiag:30859691

```

</details>

**Comment by debonte** [comment]
I'm unable to repro this. Can you provide more details?

- What version of Pylance do you have installed?
- What variable names (old / new) are you using?
- A code sample and/or video might be helpful.

**Comment by SmartLamScott** [comment]
I am using Pylance v2023.10.20.
Here's a few screenshots of the process.
1. The original code
![image](https://github.com/microsoft/pylance-release/assets/134100828/cb40e0c1-8ff1-473e-9a85-9f71e30cd788)
2. Select "old_name" on line 1, then run the Rename Symbol command, using "new_name" as the new name.
![image](https://github.com/microsoft/pylance-release/assets/134100828/9e967511-125e-48b8-9e48-103b3393a3e0)
3. After executing the command, the parentheses on line 4 are deleted.
![image](https://github.com/microsoft/pylance-release/assets/134100828/839f24d7-4254-41b7-8c9b-a7343d321449)

**Comment by debonte** [comment]
I see. Your repro steps above stated that the whitespace between `elif` and the new variable name was removed, which would result in invalid code. In your screenshots above however, and in my testing, that whitespace is unchanged. The parentheses that are removed are superfluous, so the resulting code is valid, although stylistically different. I noticed that any leading/trailing whitespace within the parentheses is also removed, but this again is superfluous.

While I agree that the parenthesis removal is unexpected for a rename operation, this seems like a low priority issue. We can handle it as an enhancement and see if other users up-vote it, indicating that they would also like to see it fixed.

Fair?

If you can provide repro steps where the resulting code is invalid, we would prioritize this more highly.

**Comment by SmartLamScott** [comment]
Sure. I'll try to find the code I was originally working on that prompted me to post the issue.

**Comment by SmartLamScott** [comment]
I can't share the original code I was working on for security purposes, but I've made an example that resembles my original issue. It seems that the problem only happens when the elif statement was immediately followed by a variable wrapped in parentheses (no spaces). I know it isn't pretty code, which is why I was going back to reformat it. Here are some screenshots.
1. Highlight the name of the variable on a line other than the elif.
![image](https://github.com/microsoft/pylance-release/assets/134100828/6d374531-148b-41ba-8597-34a51573b2ef)
2. Change the variable name to something else. Note that the extension is about to change the variable name and surrounding parenthesis on the line with the elif.
![image](https://github.com/microsoft/pylance-release/assets/134100828/6985a1f8-05fb-4e33-b6a7-b264a0a7a86b)
3. The resulting change deletes the surrounding parentheses on the line with the elif, resulting in invalid code.
![image](https://github.com/microsoft/pylance-release/assets/134100828/db646a4e-220d-4b22-bc57-67b7860abab7)

**Comment by debonte** [comment]
Great, thanks for the repro! Since this results in invalid code, it's clearly much worse. We'll take a look.

**Comment by debonte** [comment]
This will be fixed next time we pull Pyright.

**Comment by rchiodo** [comment]
This issue has been fixed in prerelease version 2023.12.104, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202312104-22-January-2024-PreRelease)
